Entity refactoring

zhil opened the pull request.

zhil committed:
#34 - refactored entities
added doctrine naming strategy
removed not used services, twig extenstion

zhil committed:
#34 - refactored load commands, services

zhil committed:
#34 - refactored dashboard, partially refactored repositories

thecatontheflat commented:
So far looks good! I will check the import functionality until Monday evening.

thecatontheflat commented:
Gives me the following error:

![pollyvoices_com_ _vagrant_vagrant___var_www_atlassian-vendor-tools-zhil_ _ssh_ _vagrant_ssh_ _211x51_and_poker__58_0_-_jira](https://cloud.githubusercontent.com/assets/727581/24351218/29c1fa34-12e5-11e7-8984-b1d749233b49.jpg)

I think in general it should be possible to allow "null" values at every path. Data provided by the API is not really consistent

zhil commented:
@thecatontheflat ok, got you

zhil committed:
#34 - refactored dashboard
moved SEN to company

zhil committed:
#34 - added company page
refactored estimated income

zhil committed:
#34 - refactored transaction section

zhil committed:
#34 - refactored licenses section

zhil committed:
#34 - fixed import for not existing billing contact

zhil commented:
@thecatontheflat I have updated PR. On my localhost all pages are working, all data seems to be the same, as on your version. But I have very small data set (just 26 licenses, 10 transactions).

I didnt refactored tests yet, probably will do it on next weekend.

Please test on your data set.

zhil committed:
#34 - removed unused code

thecatontheflat commented:
@zhil Import of licenses worked (1962 licenses; 1868 companies)!
However, importing of transactions failed:

`SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry 'AT-33760893' for key 'UNIQ_723705D12FC0CB0F'`

Could you please have a look? Let me know if you need extra data.

zhil committed:
#34 - flush transaction as soon, as we create entity

zhil committed:
#34 - added loging

zhil commented:
@thecatontheflat not sure, what could be reason.
I added flush right after creating new transaction.

Also you could enable full log to terminal using
```
./app/console app:import:transaction --log-json -vvv
```

please try again.

thecatontheflat commented:
@zhil I've commented this flush() and the logic worked:

https://github.com/zhil/atlassian-vendor-tools/blob/master/src/AppBundle/Command/ImportTransactionCommand.php#L125

I think the problem is that when a new `Company` is created and gets the `Transaction` assigned, it then cascades the empty `Transaction`. The import fails then because there are `null` fields in a freshly created object.

What do you think about removing those 2 flushes:

- https://github.com/zhil/atlassian-vendor-tools/blob/master/src/AppBundle/Command/ImportTransactionCommand.php#L106
- https://github.com/zhil/atlassian-vendor-tools/blob/master/src/AppBundle/Command/ImportTransactionCommand.php#L125

and adding one flush somewhere in the end of the loop?

zhil committed:
#34 - optimized flushing

zhil commented:
@thecatontheflat company is not created in app:import:transaction.
licenses always have company attached.
also company entity doesnt have cascades configured.

transaction need to be flushed in database as soon, as its created. Because if we will get new transaction during command execution - we may get last transaction as first on next page.

regarding line 125 - it's probably could be removed.

I have updated command, please check

thecatontheflat commented:
Sorry I was referring to getting the existing company, not creating it. The fact is that it cascades the Transaction on Company flush.

Importing works now!

I will check the overall functionality and will reply.

thecatontheflat commented:
Looks great @zhil! 
Could you fix the tests (or clean them up if redundant)?

zhil commented:
@thecatontheflat thanks

yes, I will definitely fix all test. Right now I am working on main project, plan to do it on weekend.

zhil commented:
@thecatontheflat bundles for generating doctrine fixtures require php7+
What do you think about upgrading all vendors and making requirement php7+?

thecatontheflat commented:
Sounds right to me!

zhil commented:
@thecatontheflat ok

composer have name "name": "thecatontheflat/marketplace-crm" - should I update it to thecatontheflat/atlassian-vendor-tools ?

also license is listed as proprietary, do you want to leave it?

thecatontheflat commented:
Yeah, the repo was renamed — it seems the package name was forgotten. Regarding the license — LGPL is a proper one. Thank you for the suggestions!

zhil commented:
@thecatontheflat one more question.
What should be done, if license info do not have email any more? (client decided to hide his contacts)

Probably, in "perfect world" system should process it in some way.
Not sure, if removing email from database would be correct, because in such case we will have client in MailChimp without linked company in database.
Not sure, if removing email from database + unsubscribing from MailChimp is correct - it looks too paranoic. 
Probably, behavior could be configured (do nothing / updated db / update db + remove from MailChimp)

Also how should we handle email change? Probably, having option like onCompanyEmailChange.removeFromMailchimp = true | false could have sense.

zhil committed:
upgraded vendors, set php requirement to 7.0, refactored unit tests

thecatontheflat commented:
From my use-cases I would rather prefer to have both emails (old and new) in the MailChimp. The subscriber can always remove his email from the system by unsubscribing from the list. In case if old email does not exist (bounce happens) — MailChimp will automatically remove it from the list.

But having this email actualised in our system would be a good feature!

zhil committed:
added status (exceptions, last execution time, available updates)
show status on site

zhil committed:
fixed version in travis

zhil committed:
fixed typo

zhil commented:
@thecatontheflat ok, we could probably add another issue, since email actualization would be too far from current task.

I have refactored tests. They are far from real cover, but it should be good start. 

I have also added few other small improvements
* status entity, which hold datetime for last successfull script execution, exceptions
* show warning for user if some cron script is not setuped (never launched actually)
* ability to show users, if updates available. It work in such a way
** when user install repo composer hook dump all git hashes from history in file app/cache/commits.history
** user will need to setup on cron on daily execution script ./app/console app:cron
** cron script download json file from you github.io, which have updates info in such a format "hash":"update description", like https://zhil.github.io/atlassian-vendor-tools.json#
** for each commit in update script check, if its not exists in app/cache/commits.history - add to status
** when you will release new feature - you will simply add hash of last commit to https://thecatontheflat.github.io/atlassian-vendor-tools.json with short description

It looks like this

![image](https://cloud.githubusercontent.com/assets/981783/24767878/48ae207c-1b09-11e7-8006-40eb5d5a2cd3.png)

zhil committed:
fixed typo

zhil commented:
@thecatontheflat ok, I did all, what I planned to do on this PR. Please check on your data set.

thecatontheflat commented on src/AppBundle/Command/ImportTransactionCommand.php:
```diff
@@ -0,0 +1,207 @@
+        $newCnt = 0;
+
+        foreach ($jsonTransactions as $jsonTransaction) {
+            if ($transaction = $transactionRepository->findOneBy(["transactionId"=>$jsonTransaction['transactionId']])) {
```
This should be changed to probably "find by transactionId AND pluginKey". Otherwise it skips transactions when the same company purchases multiple licenses. Example from production:

![_mysql_5_5_41-0ubuntu0_14_04_1__j-simple-edit_marketplace-crm_sale](https://cloud.githubusercontent.com/assets/727581/24828888/aa02f636-1c67-11e7-9e6b-e034b3c4167e.jpg)

Same data is missing in the current branch:
![_mysql_5_7_16-0ubuntu0_16_04_1__polly_dev_atlassian-vendor-tools_transaction_and__mysql_5_5_41-0ubuntu0_14_04_1__j-simple-edit_marketplace-crm_sale](https://cloud.githubusercontent.com/assets/727581/24828898/ca8a0b88-1c67-11e7-9b39-21fa4b86ee68.jpg)

thecatontheflat replied:
Could you please correct this logic?

zhil replied:
@thecatontheflat I see, so atlassian combine few purchases from the same vendor into single AT-XXX transaction? (which is correct, probably, that the same credit card charge)
got it, I will fix it

thecatontheflat replied:
It's not a bug 🐛  A unique transaction is identified by invoice + plugin. This API is full of mystery 😃 

Yes, I guess you're right regarding the credit card charge. Anyway, they could have provided some unique value.

zhil replied:
@thecatontheflat 
I will need to add this case to unittests, change transaction page and processing scripts. Working on another project now, will do it in nearest couple of days.

thecatontheflat replied:
Thanks a lot!

zhil replied:
@thecatontheflat for each transaction we have transactionId, addonLicenseId and hostLicenseId.
Can you please check - are they the same in this case (when the same company purchase 2 addons)?

I mean, should I change $license->addon from ManyToOne to ManyToMany?

thecatontheflat replied:
- `transactionId` — same
- `addonLicenseId` — different
- `hostLicenseId` — same

Not sure about the relations, but I trust your judgement for this case.

Attaching anonymised piece of data.


[transactions.txt](https://github.com/thecatontheflat/atlassian-vendor-tools/files/909593/transactions.txt)

zhil replied:
@thecatontheflat according to your sample data - vendorAmount is "total" for 2 plugins.
So it its impossible to calculate total sales for single plugin (because it will also include other plugins sales to the same company).

From what I see, the only one possible way - is getting transactions for all plugins one-by-one.

Can you please compare vendorAmount in 2 reports?
```
curl -u YOUR_VENDOR_EMAIL "https://marketplace.atlassian.com/rest/2/vendors/YOUR_VENDOR_ID/reporting/sales/transactions?addon=agile.retrospectives&text=33476578"

curl -u YOUR_VENDOR_EMAIL "https://marketplace.atlassian.com/rest/2/vendors/YOUR_VENDOR_ID/reporting/sales/transactions?addon=agile.estimation.3.0_private&text=33476578"
```

thecatontheflat replied:
It's not total, it's just same price (`123.75`) for each plugin. Or do I miss something?

zhil replied:
@thecatontheflat can you please check your prices for plugins?
as I see, Agile estimate is $125 for 500 users. Since you have $123.75 - probably, thats total for 2 plugins?

thecatontheflat replied:
I think I confused you with "price" and "vendor amount".

`vendorAmount` is a vendor's cut after Atlassian takes its 25% fee.
`500` users tier price is `$165`, which makes it `$123.75` after deducting Paid-via-Atlassian fee.

zhil replied:
@thecatontheflat ok, great. thanks for explanation

zhil replied:
@thecatontheflat I have a quick question

Lets say, I purchase your plugin1 on 2017-04-14, got transaction AT-001, maintenanceEndDate 2017-05-14
After 2 weeks I have installed your plugin2.
AS I understand
* on 2017-05-14 AT-001 will have only plugin1
* on  2017-06-14 AT-001 will have both plugins, so in fact I will get 6 weeks trial?

thecatontheflat replied:
I have no idea about such use-case, as I've never tried to investigate it. I know for sure that sometimes a customer might get 2 weeks of trial and sometimes 6. I would say those are edge cases. 

Normally Cloud Add-Ons trial duration is attached to the billing cycle, with some margin. For example (from my observations), if your billing cycle is 1th of the month and you start trial of an add-on at 10th, this would end the trial in less then 4 weeks (on the 1st of the next month). However, if you start your trial somewhere around 25th, you'll maybe get an extra trial week.

So regarding the described case, I **assume** that plugin2 trial will end either with plugin1, or might get extended until the next billing cycle.

zhil replied:
@thecatontheflat I have refactored system, added simple unittest for this case and changed transaction page.

Please test current version

thecatontheflat reviewed (changes requested):
Could you please have a look at the comment?

zhil commented on src/AppBundle/Command/ImportTransactionCommand.php:
```diff
@@ -0,0 +1,207 @@
+        $newCnt = 0;
+
+        foreach ($jsonTransactions as $jsonTransaction) {
+            if ($transaction = $transactionRepository->findOneBy(["transactionId"=>$jsonTransaction['transactionId']])) {
```
@thecatontheflat in your sample (first screenshot) - first 2 transactions have the same invoce ids (AT-XXXXX) - isnt it bug in current version?

zhil replied:
I supposed, that transactionId should be unique

zhil replied:
@thecatontheflat well, thats actually seems to be total.
when you have 2 plugins in the same trunsaction (lets say first have vendorAmount $100, second - 23.75) - atlassian seems to show total vendor amount for transaction - 123.75

because of it we cant separate vendor amounts for each plugin

zhil committed:
#35 - refactored system for non unique transactionId

thecatontheflat commented:
Looks super-amazing 👍 
Thank you SO MUCH!

zhil commented:
@thecatontheflat thanks :)

thank you too for cool tool, great bundle and for making them open-sources ;)
